internal/langserver/diagnostics/diagnostics.go

```diff
@@ -16,26 +16,30 @@ import (
 type diagContext struct {
 	ctx    context.Context
 	uri    lsp.DocumentURI
-	diags  hcl.Diagnostics
 	source string
+	diags  []lsp.Diagnostic
 }
 
 // Notifier is a type responsible for queueing hcl diagnostics to be converted
 // and sent to the client
 type Notifier struct {
+	logger         *log.Logger
 	sessCtx        context.Context
 	diags          chan diagContext
+	diagsCache     map[string]map[string][]lsp.Diagnostic
 	closeDiagsOnce sync.Once
 }
 
 func NewNotifier(sessCtx context.Context, logger *log.Logger) *Notifier {
-	diags := make(chan diagContext, 50)
-	go notify(diags, logger)
-	return &Notifier{sessCtx: sessCtx, diags: diags}
+	n := &Notifier{logger: logger, sessCtx: sessCtx, diags: make(chan diagContext, 50), diagsCache: make(map[string]map[string][]lsp.Diagnostic)}
+	go n.notify()
+	return n
 }
 
-// Publish accepts a map of diagnostics per file and queues them for publishing
-func (n *Notifier) Publish(ctx context.Context, rmDir string, diags map[string]hcl.Diagnostics, source string) {
+// PublishHCLDiags accepts a map of hcl diagnostics per file and queues them for publishing.
+// A dir path is passed which is joined with the filename keys of the map, to form a file URI.
+// A source string is passed and set for each diagnostic, this is typically displayed in the client UI.
+func (n *Notifier) PublishHCLDiags(ctx context.Context, dir string, diags map[string]hcl.Diagnostics, source string) {
```

Comment:
```suggestion
func (n *Notifier) PublishHCLDiags(ctx context.Context, dirPath string, diags map[string]hcl.Diagnostics, source string) {
```
I know the comment above clarifies that it's a path and not URI, but I think it could be even more obvious from the name. 😉